File: src/tier3/typescript/packages/state/src/piiAnnotation.ts
// piiAnnotation.ts — field_pii annotation の TypeScript decorator 定義 + metadata 取得
// emitPurgeEvent.ts の stripPiiFields は runtime strip だが、compile-time に annotation 漏れを検出する機構
// "experimentalDecorators": true + "emitDecoratorMetadata": true を tsconfig.json で有効化すること
// Y-tier3-4: field_pii annotation compile-time 強制（4 言語等価強度の TypeScript 実装）

// PII フィールド名を蓄積するための WeakMap（target object → Set<string> のマッピング）
// WeakMap を使うことで対象オブジェクトが GC された際にメモリリークを防止する
const _piiFieldsRegistry: WeakMap<object, Set<string>> = new WeakMap();

// PII_METADATA_KEY は reflect-metadata のキー（実験的 metadata API を使用する場合に使用する）
// ここでは WeakMap ベースの実装を primary とし、reflect-metadata は optional とする
const PII_METADATA_KEY = "k1s0:pii_fields";

// FieldPii は TypeScript PropertyDecorator を返すファクトリ関数（引数なし）
// @FieldPii() を PII を含むプロパティに付与することで annotation を宣言する
// 使用例: class Foo { @FieldPii() email: string = ""; }
// experimentalDecorators: true が tsconfig に必要
export function FieldPii(): PropertyDecorator {
  // PropertyDecorator 本体を返す（target: クラスのプロトタイプ / propertyKey: プロパティ名）
  return function (target: object, propertyKey: string | symbol): void {
    // target に対する PII フィールド名のセットを取得する（存在しない場合は新規作成する）
    let fields = _piiFieldsRegistry.get(target);
    // セットが存在しない場合は新規作成して WeakMap に登録する
    if (fields === undefined) {
      // 新規の Set を生成する
      fields = new Set<string>();
      // target の WeakMap エントリに登録する
      _piiFieldsRegistry.set(target, fields);
    }
    // プロパティ名を PII フィールドセットに追加する（symbol は toString で変換する）
    fields.add(typeof propertyKey === "symbol" ? propertyKey.toString() : propertyKey);
    // reflect-metadata が利用可能な場合はメタデータとしても記録する（optional 統合）
    if (typeof Reflect !== "undefined" && typeof Reflect.defineMetadata === "function") {
      // 既存のメタデータを取得する（なければ空配列）
      const existing: string[] =
        (Reflect.getMetadata(PII_METADATA_KEY, target) as string[] | undefined) ?? [];
      // 新しいプロパティ名を追加する
      const key = typeof propertyKey === "symbol" ? propertyKey.toString() : propertyKey;
      // 既存に含まれない場合のみ追加する（重複防止）
      if (!existing.includes(key)) {
        // メタデータを更新する
        Reflect.defineMetadata(PII_METADATA_KEY, [...existing, key], target);
      }
    }
  };
}

// getPiiFields は target オブジェクト（クラスインスタンスまたはプロトタイプ）から
// @FieldPii() が付与されたプロパティ名のリストを返す
// target: @FieldPii() を保持するクラスのプロトタイプまたはインスタンス
// 返値: PII フィールド名の配列（順序は登録順）
export function getPiiFields(target: object): string[] {
  // WeakMap から PII フィールドセットを取得する
  const fields = _piiFieldsRegistry.get(target);
  // フィールドセットが存在しない場合は空配列を返す
  if (fields === undefined) {
    // PII フィールドが未登録のため空配列を返す
    return [];
  }
  // Set を配列に変換して返す
  return Array.from(fields);
}

// getPiiFieldsFromClass は class コンストラクタからプロトタイプを解決して PII フィールド名を返す
// getPiiFields の convenience wrapper（コンストラクタを渡す場合に使用する）
// ctor: @FieldPii() を持つクラスのコンストラクタ関数
export function getPiiFieldsFromClass(ctor: new (...args: unknown[]) => object): string[] {
  // コンストラクタの prototype から PII フィールドを取得する
  return getPiiFields(ctor.prototype as object);
}

// assertNoPiiLeak は payload 内に annotation されていない PII フィールドが含まれないことをアサートする
// runtime チェック: compiletime 強制（ESLint rule）の補完として使用する
// payload: チェック対象のオブジェクト
// annotatedPiiFields: @FieldPii() で宣言された PII フィールド名のリスト
// allowedFields: PII でない許可フィールド名のリスト（payload に含まれてよいフィールド）
// 返値: 未宣言の PII フィールド名のリスト（空配列 = 問題なし）
export function assertNoPiiLeak(
  // チェック対象のオブジェクト
  payload: Record<string, unknown>,
  // @FieldPii() で宣言された PII フィールド名のリスト
  annotatedPiiFields: string[],
  // 許可フィールド名のリスト（PII でないフィールド）
  allowedFields: string[]
): string[] {
  // payload のフィールド名を取得する
  const payloadKeys = Object.keys(payload);
  // allowedFields と annotatedPiiFields のユニオンセットを生成する
  const knownFields = new Set<string>([...allowedFields, ...annotatedPiiFields]);
  // payload に含まれる未知のフィールド名を返す（PII 漏洩の可能性があるフィールド）
  return payloadKeys.filter((k) => !knownFields.has(k));
}

// PiiAnnotatedClass は @FieldPii() を使用するクラスに implements させる interface
// compile-time の型チェックでアノテーション忘れを防止する（optional 使用）
export interface PiiAnnotatedClass {
  // __piiAnnotated フラグ（存在確認用の marker property）
  readonly __piiAnnotated: true;
}

File: src/tier3/go/state/pii_annotation.go
// pii_annotation.go — field_pii annotation の Go 実装（reflect タグベース + compile-time 強制）
// struct tag `pii:"true"` を使って PII フィールドを宣言し、
// GetPiiFields で reflect を使って PII フィールド名を抽出する
// Y-tier3-4: field_pii annotation compile-time 強制（4 言語等価強度の Go 実装）
package state

import (
	// reflect パッケージ: struct tag の読み取りに使用する
	"reflect"
)

// PiiField は struct tag `pii:"true"` を持つフィールドのマーカー型（型エイリアス）
// Go では decorator がないため struct tag で宣言する（`pii:"true"` タグが SoT）
// 使用例: type Foo struct { Email string `pii:"true"` }
type PiiField = struct{}

// PiiTagKey は struct tag のキー文字列（`pii:"true"` の "pii" 部分）
// 全コードで統一的に使用する定数として定義する
const PiiTagKey = "pii"

// PiiTagValue は PII フィールドを宣言する tag 値（`pii:"true"` の "true" 部分）
// "true" 以外の値は PII フィールドとして認識しない
const PiiTagValue = "true"

// PiiAnnotated は PII フィールドを持つ struct が実装するインタフェース
// GetPiiFieldNames: 実装クラスが PII フィールド名のスライスを返す
// このインタフェースの実装は自動ではなく手動（将来は go:generate で生成する）
type PiiAnnotated interface {
	// GetPiiFieldNames は PII フィールド名のスライスを返す
	GetPiiFieldNames() []string
}

// GetPiiFields は v の reflect.Type を走査して `pii:"true"` tag を持つフィールド名を返す
// v: PII フィールドを持つ struct の interface{} 値（ポインタ / 値 いずれも可）
// 返値: `pii:"true"` tag を持つフィールド名のスライス（順序は struct 定義順）
func GetPiiFields(v interface{}) []string {
	// v の reflect.Type を取得する（ポインタの場合は Elem() で実体を取得する）
	t := reflect.TypeOf(v)
	// t が nil の場合は空スライスを返す（nil 引数への安全なフォールバック）
	if t == nil {
		// nil 引数は空スライスを返す
		return []string{}
	}
	// ポインタの場合は Elem() で実体の型を取得する
	if t.Kind() == reflect.Ptr {
		// ポインタを Dereference して実体の型を取得する
		t = t.Elem()
	}
	// struct でない場合は空スライスを返す（struct 以外は tag を持てない）
	if t.Kind() != reflect.Struct {
		// struct 以外は空スライスを返す
		return []string{}
	}
	// PII フィールド名を格納するスライスを生成する
	piiFields := make([]string, 0, t.NumField())
	// struct の全フィールドを走査する
	for i := 0; i < t.NumField(); i++ {
		// i 番目のフィールドを取得する
		field := t.Field(i)
		// struct tag から "pii" キーの値を取得する（`pii:"true"` の場合は "true" が返る）
		tagVal := field.Tag.Get(PiiTagKey)
		// tag 値が PiiTagValue（"true"）の場合は PII フィールドとして追加する
		if tagVal == PiiTagValue {
			// フィールド名を PII フィールドスライスに追加する
			piiFields = append(piiFields, field.Name)
		}
	}
	// PII フィールド名のスライスを返す（空スライスは []string{} になる）
	return piiFields
}

// HasPiiFields は v が `pii:"true"` tag を持つフィールドを 1 件以上持つかどうかを返す
// GetPiiFields の convenience wrapper（存在確認のみが目的の場合に使用する）
func HasPiiFields(v interface{}) bool {
	// GetPiiFields を呼び出してフィールド数を確認する
	return len(GetPiiFields(v)) > 0
}

// ValidatePiiStripped は payload map から PII フィールドが除去済みかを検証する
// GetPiiFields が返すフィールド名が payload に含まれていないことを確認する
// v: PII フィールドの宣言元 struct の interface{} 値
// payload: 送信対象の map（PII フィールドが strip 済みであることを確認する）
// 返値: payload に残存する PII フィールド名のスライス（空スライス = strip 完了）
func ValidatePiiStripped(v interface{}, payload map[string]interface{}) []string {
	// v から PII フィールド名を取得する
	piiFields := GetPiiFields(v)
	// payload に残存する PII フィールドを格納するスライスを生成する
	remaining := make([]string, 0)
	// PII フィールドを走査して payload に残存するか確認する
	for _, field := range piiFields {
		// payload に PII フィールドが存在する場合は残存リストに追加する
		if _, exists := payload[field]; exists {
			// strip 漏れの PII フィールドを追加する
			remaining = append(remaining, field)
		}
	}
	// 残存する PII フィールド名のスライスを返す（空スライス = strip 完了）
	return remaining
}
